micro-rdk/src/esp32/camera.rs

```diff
@@ -1,117 +1,265 @@
-        }
+            last_grab: Instant::now(),
+        };
+
+        let ret =
+            (unsafe { esp_camera_init(&cam.config) }) as crate::esp32::esp_idf_svc::sys::esp_err_t;
+        EspError::convert(ret).map_err(|e| CameraError::InitError(e.into()))?;
+        Ok(Arc::new(Mutex::new(cam)))
     }
-    pub fn debug_print_fb(&self, frame: &Option<*mut crate::esp32::esp_idf_svc::sys::camera_fb_t>) {
-        if let Some(ptr) = frame {
-            let ptr = ptr as &*mut crate::esp32::esp_idf_svc::sys::camera_fb_t;
-            unsafe {
-                println!();
-                info!("camera buf size {}", (*(*ptr)).len);
-                println!();
-                for i in 0..(*(*ptr)).len {
-                    print!("{:02X}", (*(*(*ptr)).buf.offset(i as isize)));
-                    if i > 0 && i % 80 == 0 {
-                        println!();
-                    }
-                }
-            }
-        }
+
+    fn get_frame(&mut self) -> Result<Esp32CameraFrameBuffer, CameraError> {
+        self.last_grab = Instant::now();
+        let frame = Esp32CameraFrameBuffer::get().ok_or_else(|| CameraError::FailedToGetImage)?;
+        Ok(frame)
     }
 }
+
 impl Camera for Esp32Camera {
-    fn get_frame(&mut self, mut buffer: BytesMut) -> Result<BytesMut, CameraError> {
-        if let Some(ptr) = self.get_cam_frame() {
-            let buf = unsafe {
-                let buf = (*ptr).buf;
-                let len = (*ptr).len as usize;
-                core::slice::from_raw_parts(buf, len)
-            };
-            if buf.len() > buffer.capacity() {
-                self.return_cam_frame(Some(ptr));
-                return Err(CameraError::CameraFrameTooBig);
+    fn get_image(&mut self, mut buffer: BytesMut) -> Result<BytesMut, CameraError> {
+        if let Ok(frame) = self.get_frame() {
+            if frame.len() > buffer.capacity() {
+                return Err(CameraError::ImageTooBig);
             }
-            let bytes = Bytes::from(buf);
+            self.last_grab = Instant::now();
             let msg = camera::v1::GetImageResponse {
                 mime_type: "image/jpeg".to_string(),
-                image: bytes,
+                image: unsafe { frame.as_bytes() },
             };
+            // safety: message must be encoded before the frame is dropped from scope
             msg.encode(&mut buffer).unwrap();
-            self.return_cam_frame(Some(ptr));
+
             return Ok(buffer);
         }
-        Err(CameraError::CameraCouldntGetFrame)
+        Err(CameraError::FailedToGetImage)
+    }
+
+    fn render_frame(&mut self, mut buffer: BytesMut) -> Result<BytesMut, CameraError> {
+        if let Ok(frame) = self.get_frame() {
+            if frame.len() > buffer.capacity() {
+                return Err(CameraError::ImageTooBig);
+            }
+            let msg = HttpBody {
+                content_type: "image/jpeg".to_string(),
```

Comment:
does the format work even if the camera is configured to capture something different than JPEG? If not we might want to commit to JPEG for this version of the driver

Reply:
committing to jpeg for this version